pkgs/by-name/co/comfyui/package.nix

```diff
@@ -0,0 +1,164 @@
+  userPath ? "/var/lib/comfyui/user",
+  customNodes ? [ ],
+  models ? {
+    checkpoints = { };
+    clip = { };
+    clip_vision = { };
+    configs = { };
+    controlnet = { };
+    embeddings = { };
+    upscale_modules = { };
+    vae = { };
+    vae_approx = { };
+  },
+}:
+
+let
+  config-data = {
+    comfyui = {
+      base_path = modelsPath;
+      checkpoints = "${modelsPath}/checkpoints";
+      clip = "${modelsPath}/clip";
+      clip_vision = "${modelsPath}/clip_vision";
+      configs = "${modelsPath}/configs";
+      controlnet = "${modelsPath}/controlnet";
+      diffusion_models = "${modelsPath}/diffusion_models";
+      embeddings = "${modelsPath}/embeddings";
+      loras = "${modelsPath}/loras";
+      style_models = "${modelsPath}/style_models";
+      text_encoders = "${modelsPath}/text_encoders";
+      upscale_models = "${modelsPath}/upscale_models";
+      vae = "${modelsPath}/vae";
+      vae_approx = "${modelsPath}/vae_approx";
+    };
+  };
+
+  modelPathsFile = writeTextFile {
+    name = "extra_model_paths.yaml";
+    text = (lib.generators.toYAML { } config-data);
+  };
+
+  pythonEnv = (
+    python312.withPackages (
+      ps:
+      with ps;
+      [
+        comfyui-frontend-package
+        comfyui-workflow-templates
+        torch
+        torchsde
+        torchvision
+        torchaudio
+        numpy
+        einops
+        transformers
+        tokenizers
+        sentencepiece
+        safetensors
+        aiohttp
+        yarl
+        pyyaml
+        pillow
+        scipy
+        tqdm
+        psutil
+
+        kornia
+        spandrel
+        av
+        pydantic
+      ]
+      ++ (builtins.concatMap (node: node.dependencies) customNodes)
+    )
+  );
+
+  executable = writers.writeDashBin "comfyui" ''
+    cd $out && \
+    ${pythonEnv}/bin/python comfyui \
+      --input-directory ${inputPath} \
+      --output-directory ${outputPath} \
+      --extra-model-paths-config ${modelPathsFile} \
+      --temp-directory ${tempPath} \
+      "$@"
+  '';
+
+  customNodesCollection = (
+    linkFarm "comfyui-custom-nodes" (
+      builtins.map (pkg: {
+        name = pkg.pname;
+        path = pkg;
+      }) customNodes
+    )
+  );
+in
+stdenv.mkDerivation rec {
+  pname = "comfyui";
+  version = "unstable-2025-04-26";
+
+  src = fetchFromGitHub {
+    owner = "comfyanonymous";
+    repo = "ComfyUI";
+    rev = "b685b8a4e098237919adae580eb29e8d861b738f";
+    hash = "sha256-OtTvyqiz2Ba7HViW2MxC1hFulSWPuQaCADeQflr80Ik=";
+  };
+
+  installPhase = ''
+    runHook preInstall
+    echo "Preparing bin folder"
+    mkdir -p $out/bin/
+    echo "Copying comfyui files"
+    # These copies everything over but test/ci/github directories.  But it's not
+    # very future-proof.  This can lead to errors such as "ModuleNotFoundError:
+    # No module named 'app'" when new directories get added (which has happened
+    # at least once).  Investigate if we can just copy everything.
+    cp -r $src/app $out/
```

Comment:
cp can copy multiple directories at once

Reply:
I guess I can put it in one long line if you want but imo it's far more readable this way

Reply:
yeah, not that important. I was trying to condense the bigger script block a bit but one one very long line is also not great. Maybe a wildcard but details in the end.